server/src/main/java/org/elasticsearch/action/ingest/ReservedPipelineAction.java

```diff
@@ -0,0 +1,170 @@
+        final IngestMetadata currentIndexMetadata = state.metadata().custom(IngestMetadata.TYPE);
+
+        var updatedIngestMetadata = task.execute(currentIndexMetadata, allIndexMetadata);
+        return state.copyAndUpdateMetadata(b -> b.putCustom(IngestMetadata.TYPE, updatedIngestMetadata));
+    }
+
+    @Override
+    public TransformState transform(Object source, TransformState prevState) throws Exception {
+        @SuppressWarnings("unchecked")
+        var requests = prepare((ReservedPipelinesData) source);
+
+        ClusterState state = prevState.state();
+
+        for (var request : requests) {
+            var nopUpdate = ingestService.isNoOpPipelineUpdate(state, request, new ActionListener<>() {
+                @Override
+                public void onResponse(AcknowledgedResponse acknowledgedResponse) {}
+
+                @Override
+                public void onFailure(Exception e) {
+                    throw new IllegalStateException(e.getMessage(), e);
+                }
+            });
+
+            if (nopUpdate) {
+                continue;
+            }
+
+            var task = new IngestService.PutPipelineClusterStateUpdateTask(request);
+            state = wrapIngestTaskExecute(task, state);
+        }
+
+        Set<String> entities = requests.stream().map(r -> r.getId()).collect(Collectors.toSet());
+
+        Set<String> toDelete = new HashSet<>(prevState.keys());
+        toDelete.removeAll(entities);
+
+        for (var pipelineToDelete : toDelete) {
+            var task = new IngestService.DeletePipelineClusterStateUpdateTask(pipelineToDelete);
+            state = wrapIngestTaskExecute(task, state);
+        }
+
+        return new TransformState(state, entities);
+    }
+
+    @Override
+    public ReservedPipelinesData fromXContent(XContentParser parser) throws IOException {
+        List<PutPipelineRequest> result = new ArrayList<>();
+
+        Map<String, ?> source = parser.map();
+
+        for (String id : source.keySet()) {
+            @SuppressWarnings("unchecked")
+            Map<String, ?> content = (Map<String, ?>) source.get(id);
+            try (XContentBuilder builder = XContentFactory.contentBuilder(XContentType.JSON)) {
+                builder.map(content);
+                result.add(new PutPipelineRequest(id, BytesReference.bytes(builder), XContentType.JSON));
+            } catch (Exception e) {
+                throw new ElasticsearchGenerationException("Failed to generate [" + source + "]", e);
+            }
+        }
+
+        return new ReservedPipelinesData(result, null);
+    }
+
+    @Override
+    public void preTransform(Object parsedContent, ActionListener<Tuple<String, ?>> listener) {
+        NodesInfoRequest nodesInfoRequest = new NodesInfoRequest();
+        nodesInfoRequest.clear();
+        nodesInfoRequest.addMetric(NodesInfoRequest.Metric.INGEST.metricName());
+        nodeClient.admin().cluster().nodesInfo(nodesInfoRequest, new ActionListener<>() {
```

Comment:
This makes me uneasy, what happens if the cluster is overloaded and this request times out? I think then we cancel all file-based changes right? (because the error will be thrown from the pre-processing if I understand correctly?)

Should this timeout be changed or configurable?